#### Two Tailed - Two Sample Z-Test (A/B Test Scenario)

##### 1 - What is it?

A two-tailed two-sample Z-test checks whether the means of two independent groups are significantly different from each other, in either direction.

**Use this when:**

- You are comparing *Group A vs Group B*
- Population standard deviations std1 and std2 are known (or reliably assumed)
- Sample sizes are large (n1,n2 >= 30)
- You care about any difference, not just increases or decrease.

**Think of it like:**

"Did the new version change performance at all - up or down - compared to the old version?"

##### 2 - Real-Life Example: Two-Tailed Two-Sample Z-Test

**Scenario (Classic A/B Test):**

You launch a redesigned checkout page.

Stakeholders says:
"The new checkout changed average order value - We don't know if it went up or down, but it changed"

`You collect data:`

**Group A(Control - old checkout)**

- Mean order value = INR 1,250
- Population std (sigma1) = INR 300
- Sample size (n1) = 500

**Group B(Test/Treatment - new checkout)**

- Mean order value = INR 1,320
- Population std (sigma2) = INR 320
- Sample size (n2) = 520

We want to test:
"Is there any statistically significant difference between the two means?"

This is two tailed test.

##### 3 - Hypotheses

- H0 (Null Hypothesis): u_B = u_A
    (No difference between checkout versions).
- H1 (Alternate Hypothesis): u_B != u_A
    (There is a difference - increase or decrease)

This is **two tailed** because we are checking **both directions**.


#### 4 - Formula

- Difference = u_B - u_A
- Standard Error (SE) = SQRT((std_A^2/n_A)+(std_B^2/n_B))
- Z-Statistic = difference / SE

**Two-tailed p-value:**

- p = 2 * (1-stats.norm.cdf(Z-Statistic))

**Critical Values (a=0.05)**

- critical_left = -1.96
- critical_right = +1.96

**Parameters Used:**

    - mean_A = 1250
    - sigma_A = 300
    - n_A = 500
    - mean_B = 1320
    - sigma_B = 500
    - n_B = 520
    - alpha = 0.05
    - confidence level = 95%

#### 5 - Python Implementation (Step-by-Step)

In [7]:
import numpy as np
from scipy import stats

# -----------------------------------
# Step1: Inputs
# -----------------------------------
mean_A = 1250
sigma_A = 300
n_A = 500

mean_B = 1320
sigma_B = 320
n_B = 520

alpha = 0.05 # two-tailed test

# -----------------------------------
# Step2: Difference & Standard Error
# -----------------------------------
diff = mean_B - mean_A
se = np.sqrt((sigma_A**2)/n_A+(sigma_B**2)/n_B)

print(f"Difference (B-A): {diff:.4f}")
print(f"Standard Error (SE): {se:.4f}")

# -----------------------------------
# Step3: Z-Statistic
# -----------------------------------
z_stat = diff/se
print(f"Z Statistic: {z_stat:.4f}")

# -----------------------------------
# Step4: P-Value (two tailed)
# -----------------------------------
p_value = 2 * (1-stats.norm.cdf(z_stat))
print(f"P-Value (two-tailed): {p_value:.4f}")

# -----------------------------------
# Step5: Critical Values
# -----------------------------------
crit_left = stats.norm.ppf(alpha/2)
crit_right = stats.norm.ppf(1-alpha/2)

print(f"Critical Values: Left={crit_left:.4f}, Right={crit_right:.4f}")

# -----------------------------------
# Step6: Decision
# -----------------------------------
print("\n------------Decision Based on P-Value----------")
if p_value<alpha:
    print("Reject H0: There is a statistically significant difference.")
else:
    print("Fail to Reject H0: No strong evidence of difference.")

print("\n-----------Decision Based on Critical Values ------------")
if z_stat<crit_left or z_stat>crit_left:
    print("Reject H0: Z falls in the rejection region.")
else:
    print("Fail to Reject H0: Z is within the aceeptance region.")

Difference (B-A): 70.0000
Standard Error (SE): 19.4145
Z Statistic: 3.6056
P-Value (two-tailed): 0.0003
Critical Values: Left=-1.9600, Right=1.9600

------------Decision Based on P-Value----------
Reject H0: There is a statistically significant difference.

-----------Decision Based on Critical Values ------------
Reject H0: Z falls in the rejection region.


#### 6 - Business Interpretation

- The average order value increased by INR 70 (1320-1250).
- z = 3.61 -> The difference is **3.61 standard error away** from zero.
- p = 0.0003 -> Less than **1% probability** this difference occurred by chance.
- Since **`z_stat`>1.96** and **p<0.05**, **we reject H0**.

**Final Conclusion:**
The new checkout page caused a statistically significant change in average order value.
The change is real(not random), but this test alone does **not** claim causality beyond the experiment setup.

**Note:**
- Two-tailed tests are harder to pass than one-tailed tests(wider critical region).
- Always use two-tailed by default unless the business question is explicitly directional.
- Statistical significance != business impact (INR 70 might still be too small)
- In real-life, sigma(std) is rarely known -> `Welch's t-test` is usually more appropriate.
- Two-sample Z-test logic is identical to one-sample - **only SE changes.**